File: OpenStreetMap/Shared/VoiceAnnouncement.m
//
//  VoiceAnnouncement.m
//  Go Map!!
//
//  Created by Bryce on 10/26/15.
//  Copyright © 2015 Bryce. All rights reserved.
//

#import "BingMapsGeometry.h"
#import "EditorMapLayer.h"
#import "MapView.h"
#import "OsmMapData.h"
#import "OsmObjects.h"
#import "VoiceAnnouncement.h"


#import <AVFoundation/AVFoundation.h>


static inline OSMPoint OSMPointFromCoordinate( CLLocationCoordinate2D coord )
{
	OSMPoint point = { coord.longitude, coord.latitude };
	return point;
}

@implementation VoiceAnnouncement

-(instancetype)init
{
	self = [super init];
	if ( self ) {
		_utteranceMap = [NSMapTable mapTableWithKeyOptions:NSPointerFunctionsOpaquePersonality valueOptions:NSPointerFunctionsOpaquePersonality];

		_buildings			= NO;
		_addresses			= NO;
		_streets			= YES;
		_shopsAndAmenities	= YES;

		_enabled			= YES;
	}
	return self;
}

-(void)say:(NSString *)text forObject:(OsmBaseObject *)object
{
	if ( _synthesizer == nil ) {
		_synthesizer = [[AVSpeechSynthesizer alloc] init];
		_synthesizer.delegate = self;
	}

	if ( object && _isNewUpdate ) {
		_isNewUpdate = NO;
		[self say:@"update" forObject:nil];
	}

	AVSpeechUtterance * utterance = [AVSpeechUtterance speechUtteranceWithString:text];
	[_synthesizer speakUtterance:utterance];

	[_utteranceMap setObject:object forKey:utterance];
}


-(void)removeAll
{
	[_synthesizer stopSpeakingAtBoundary:AVSpeechBoundaryWord];
	[_utteranceMap removeAllObjects];
}

-(void)setEnabled:(BOOL)enabled
{
	if ( enabled != _enabled ) {
		_enabled = enabled;
		if ( !enabled ) {
			[self removeAll];
		}
	}
}


-(void)announceForLocation:(CLLocationCoordinate2D)coord
{
	if ( !self.enabled )
		return;

	_isNewUpdate = YES;

	NSMutableArray * a = [NSMutableArray arrayWithCapacity:100];

	CGPoint metersPerDegree = { MetersPerDegreeLongitude(coord.latitude), MetersPerDegreeLatitude(coord.latitude) };
	if ( _previousCoord.latitude == 0 && _previousCoord.longitude == 0 )
		_previousCoord = coord;
	OSMRect box = { MIN(_previousCoord.longitude,coord.longitude), MIN(_previousCoord.latitude,coord.latitude), fabs(_previousCoord.longitude-coord.longitude), fabs(_previousCoord.latitude-coord.latitude) };
	box.origin.x -= _radius/metersPerDegree.x;
	box.origin.y -= _radius/metersPerDegree.y;
	box.size.width  += 2*_radius/metersPerDegree.x;
	box.size.height += 2*_radius/metersPerDegree.y;

	[self.mapView.editorLayer.mapData enumerateObjectsInRegion:box block:^(OsmBaseObject *obj) {
		if ( obj.deleted )
			return;
		if ( !obj.hasInterestingTags )
			return;
		// make sure it is within distance
		double dist = [obj distanceToLineSegment:OSMPointFromCoordinate(_previousCoord) point:OSMPointFromCoordinate(coord)];
		if ( dist < self.radius ) {
			[a addObject:@[ @(dist), obj ]];
		}
	}];

	// sort by distance
	[a sortedArrayUsingComparator:^NSComparisonResult(NSArray * obj1, NSArray * obj2) {
		double d1 = [obj1[0] doubleValue];
		double d2 = [obj2[0] doubleValue];
		return d1 < d2 ? NSOrderedAscending : d1 > d2 ? NSOrderedDescending : NSOrderedSame;
	}];

	NSDate * now = [NSDate new];
	NSMutableDictionary * currentObjects = [NSMutableDictionary new];
	OsmWay * closestHighwayWay  = nil;
	double	 closestHighwayDist = 1000000.0;
	OsmWay * newCurrentHighway = nil;
	for ( NSArray * item in a ) {
		OsmBaseObject * object	 = item[1];

		// track highway we're closest to
		if ( object.isWay && object.tags[@"highway"] ) {
			double	distance = [item[0] doubleValue];
			if ( distance < closestHighwayDist ) {
				closestHighwayDist = distance;
				closestHighwayWay  = object.isWay;
			}
		}
	}
	if ( closestHighwayWay && closestHighwayWay == _previousClosestHighway ) {
		if ( closestHighwayWay != _currentHighway ) {
			_currentHighway = closestHighwayWay;
			newCurrentHighway = _currentHighway;
		}
	}
	_previousClosestHighway = closestHighwayWay;

	for ( NSArray * item in a ) {
		OsmBaseObject * object	 = item[1];

		// if we've recently announced object then don't announce again
		NSNumber * ident = @(object.extendedIdentifier);
		currentObjects[ident] = now;
		if ( _previousObjects[ident] && object != newCurrentHighway )
			continue;

		if ( _buildings && object.tags[@"building"] ) {
			NSString * building = object.tags[@"building"];
			if ( [building isEqualToString:@"yes"] )
				building = @"";
			NSString * text = [NSString stringWithFormat:@"building %@",building];
			[self say:text forObject:object];
		}

		if ( _addresses && object.tags[@"addr:housenumber"] ) {
			NSString * number = object.tags[@"addr:housenumber"];
			NSString * street = object.tags[@"addr:street"];
			NSString * text = [NSString stringWithFormat:@"%@ number %@",street, number];
			[self say:text forObject:object];
		}
		
		if ( _streets && object.isWay && object.tags[@"highway"] ) {
			NSString * type = object.tags[@"highway"];
			NSString * name = object.tags[@"name"];
			if ( name == nil )
				name = object.tags[@"ref"];
			if ( [type isEqualToString:@"service"] && name == nil && object != newCurrentHighway ) {
				// skip
			} else {
				NSString * text = name ?: type;
				if ( object == newCurrentHighway )
					text = [NSString stringWithFormat:@"Now following %@",text];
				[self say:text forObject:object];
			}
		}
	
		if ( _shopsAndAmenities ) {
			if ( object.tags[@"shop"] || object.tags[@"amenity"] ) {
				NSString * text = object.friendlyDescription;
				[self say:text forObject:object];
			}
		}
	}

	_previousObjects = currentObjects;
	_previousCoord	= coord;
}

#pragma mark delegate

- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didStartSpeechUtterance:(AVSpeechUtterance *)utterance
{
	OsmBaseObject * object = [_utteranceMap objectForKey:utterance];
	_mapView.editorLayer.selectedNode		= object.isNode;
	_mapView.editorLayer.selectedWay		= object.isWay;
	_mapView.editorLayer.selectedRelation	= object.isRelation;
	[_utteranceMap removeObjectForKey:utterance];
}

- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didFinishSpeechUtterance:(AVSpeechUtterance *)utterance
{
	_mapView.editorLayer.selectedNode		= nil;
	_mapView.editorLayer.selectedWay		= nil;
	_mapView.editorLayer.selectedRelation	= nil;
}

- (void)speechSynthesizer:(AVSpeechSynthesizer *)synthesizer didCancelSpeechUtterance:(AVSpeechUtterance *)utterance
{
	[self speechSynthesizer:synthesizer didFinishSpeechUtterance:utterance];
}


@end

File: src/Shared/VoiceAnnouncement.swift
//
//  VoiceAnnouncement.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 10/26/15.
//  Copyright © 2015 Bryce Cogswell. All rights reserved.
//

import AVFoundation
import Foundation
import CoreLocation

class VoiceAnnouncement: NSObject, AVSpeechSynthesizerDelegate {
    var synthesizer: AVSpeechSynthesizer?
    var previousObjects = [OsmExtendedIdentifier : Date]()
	var previousCoord = LatLon.zero
	var currentHighway: OsmWay? = nil
    var previousClosestHighway: OsmWay? = nil
    var utteranceMap = NSMapTable<AVSpeechUtterance, OsmBaseObject>(keyOptions: .opaquePersonality, valueOptions: .opaquePersonality)

    var isNewUpdate = false

    var mapView: MapView!
    var radius = 0.0
    var buildings = false
    var streets = false
    var addresses = false
    var shopsAndAmenities = false

    private var _enabled = false
    var enabled: Bool {
        get {
            _enabled
        }
        set(enabled) {
            if enabled != _enabled {
                _enabled = enabled
                if !enabled {
                    removeAll()
                }
            }
        }
    }

    override init() {
        super.init()

        buildings = false
        addresses = false
        streets = true
        shopsAndAmenities = true

        enabled = true
    }

    func say(_ text: String, for object: OsmBaseObject?) {
        if synthesizer == nil {
            synthesizer = AVSpeechSynthesizer()
            synthesizer?.delegate = self
        }

        if object != nil && isNewUpdate {
            isNewUpdate = false
            say("update", for: nil)
        }

        let utterance = AVSpeechUtterance(string: text)
        synthesizer?.speak(utterance)

        utteranceMap.setObject(object, forKey: utterance)
    }

    func removeAll() {
        synthesizer?.stopSpeaking(at: .word)
        utteranceMap.removeAllObjects()
    }

    func announce(forLocation coord: LatLon) {
		guard let mapView = mapView else { return }
        if !enabled {
            return
        }

        isNewUpdate = true

		let metersPerDegree = MetersPerDegreeAt(latitude:coord.latitude)
		if previousCoord.latitude == 0.0 && previousCoord.longitude == 0.0 {
            previousCoord = coord
        }
        
//        OSMRect.init(origin: OSMPoint(x: min(previousCoord?.longitude, coord.longitude), y: min(previousCoord?.latitude, coord.latitude)), size: OSMSize(width: abs(), height: ))
		var box = OSMRect(origin:OSMPoint(x:min(previousCoord.longitude, coord.longitude),
										  y:min(previousCoord.latitude, coord.latitude)),
						  size:OSMSize(width: abs(previousCoord.longitude - coord.longitude),
									   height: abs(previousCoord.latitude - coord.latitude)))
        box.origin.x -= radius / Double(metersPerDegree.x)
        box.origin.y -= radius / Double(metersPerDegree.y)
        box.size.width += 2 * radius / Double(metersPerDegree.x)
        box.size.height += 2 * radius / Double(metersPerDegree.y)

		var a = [(Double,OsmBaseObject)]()

		mapView.editorLayer.mapData.enumerateObjects(inRegion: box, block: { obj in
            if obj.deleted {
                return
            }
            if !obj.hasInterestingTags() {
                return
            }
            // make sure it is within distance
			let dist = obj.distance(toLineSegment: OSMPointFromCoordinate(self.previousCoord),
									point: OSMPointFromCoordinate(coord))

			if dist < self.radius {
				if let currentHighway = self.currentHighway,
				   let way = obj.isWay(),
				   obj.tags["highway"] != nil,
				   !currentHighway.sharesNodes(with: way)
				{
					return // only announce ways connected to current way
                }
                a.append((dist, obj))
            }
        })

        // sort by distance
		a.sort { obj1,obj2 in return obj1.0 < obj2.0 }

        let now = Date()
        var currentObjects: [OsmExtendedIdentifier : Date] = [:]
        var closestHighwayWay: OsmWay? = nil
        var closestHighwayDist = 1000000.0
        var newCurrentHighway: OsmWay? = nil
        for (distance,object) in a {

            // track highway we're closest to
            if object.isWay() != nil && object.tags["highway"] != nil {
				if distance < closestHighwayDist {
                    closestHighwayDist = distance
                    closestHighwayWay = object.isWay()
                }
            }
        }
        if closestHighwayWay != nil && closestHighwayWay == previousClosestHighway {
            if closestHighwayWay != currentHighway {
                currentHighway = closestHighwayWay
                newCurrentHighway = currentHighway
            }
        }
        previousClosestHighway = closestHighwayWay

        for (_,object) in a {

            // if we've recently announced object then don't announce again
            let ident = object.extendedIdentifier
            currentObjects[ident] = now
			if previousObjects[ident] != nil && object != newCurrentHighway {
                continue
            }

            if buildings && object.tags["building"] != nil {
				var building = object.tags["building"] ?? ""
                if building == "yes" {
                    building = ""
                }
                let text = "building \(building)"
                say(text, for: object)
            }

            if addresses,
			   let number = object.tags["addr:housenumber"]
			{
                let street = object.tags["addr:street"]
                let text = "\(street ?? "") number \(number)"
                say(text, for: object)
            }

            if streets,
			   object.isWay() != nil,
			   let type = object.tags["highway"]
			{
                let name = object.tags["name"] ?? object.tags["ref"]
                if type == "service" && name == nil && object != newCurrentHighway {
                    // skip
                } else {
                    var text = name ?? type
                    if object == newCurrentHighway {
						text = "Now following \(text)"
                    }
                    say(text, for: object)
                }
            }

            if shopsAndAmenities {
                if object.tags["shop"] != nil || object.tags["amenity"] != nil {
                    let text = object.friendlyDescription()
                    say(text, for: object)
                }
            }
        }

        previousObjects = currentObjects
        previousCoord = coord
    }

    // MARK: delegate

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance) {
		let object = utteranceMap.object(forKey: utterance)
        mapView.editorLayer.selectedNode = object?.isNode()
        mapView.editorLayer.selectedWay = object?.isWay()
        mapView.editorLayer.selectedRelation = object?.isRelation()
        utteranceMap.removeObject(forKey: utterance)
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance) {
        mapView.editorLayer.selectedNode = nil
        mapView.editorLayer.selectedWay = nil
        mapView.editorLayer.selectedRelation = nil
    }

    func speechSynthesizer(_ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance) {
        speechSynthesizer(synthesizer, didFinish: utterance)
    }
}

@inline(__always) private func OSMPointFromCoordinate(_ coord: LatLon) -> OSMPoint {
    let point = OSMPoint(x: coord.longitude, y: coord.latitude)
    return point
}
